crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+            },
+            |remaining| self.store.multi_get_effects(remaining.iter()),
+        )
+    }
+
+    fn notify_read_executed_effects_digests<'a>(
+        &'a self,
+        digests: &'a [TransactionDigest],
+    ) -> BoxFuture<'a, SuiResult<Vec<TransactionEffectsDigest>>> {
+        async move {
+            let registrations = self
+                .executed_effects_digests_notify_read
+                .register_all(digests);
+
+            let executed_effects_digests = self.multi_get_executed_effects_digests(digests)?;
+
+            let results = executed_effects_digests
+                .into_iter()
+                .zip(registrations)
+                .map(|(a, r)| match a {
+                    // Note that Some() clause also drops registration that is already fulfilled
+                    Some(ready) => Either::Left(futures::future::ready(ready)),
+                    None => Either::Right(r),
+                });
+
+            Ok(join_all(results).await)
+        }
+        .boxed()
+    }
+
+    fn multi_get_events(
+        &self,
+        event_digests: &[TransactionEventsDigest],
+    ) -> SuiResult<Vec<Option<TransactionEvents>>> {
+        do_fallback_lookup(
+            event_digests,
+            |digest| {
+                if let Some(events) = self.dirty.transaction_events.get(digest) {
+                    CacheResult::Hit(Some(events.1.clone()))
+                } else {
+                    CacheResult::Miss
+                }
+            },
+            |digests| self.store.multi_get_events(digests),
+        )
+    }
+
+    fn get_sui_system_state_object_unsafe(&self) -> SuiResult<SuiSystemState> {
+        get_sui_system_state(self)
+    }
+
+    fn get_marker_value(
+        &self,
+        object_id: &ObjectID,
+        version: &SequenceNumber,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<MarkerValue>> {
+        if let Some(markers) = self.dirty.markers.get(&(epoch_id, *object_id)) {
+            if let Some(marker) = markers.get(version) {
+                return Ok(Some(*marker));
+            }
+        }
+
+        if let Some(markers) = self.cached.marker_cache.get(&(epoch_id, *object_id)) {
+            if let Some(marker) = markers.lock().get(version) {
+                return Ok(Some(*marker));
+            }
+        }
+
+        // NOTE: we cannot insert this marker into the cache without breaking the
+        // contiguous version property of the cache.
+        self.store.get_marker_value(object_id, version, epoch_id)
+    }
+
+    fn get_latest_marker(
+        &self,
+        object_id: &ObjectID,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<(SequenceNumber, MarkerValue)>> {
+        if let Some(markers) = self.dirty.markers.get(&(epoch_id, *object_id)) {
+            if let Some((k, v)) = markers.get_last() {
+                return Ok(Some((*k, *v)));
+            }
+        }
+
+        if let Some(markers) = self.cached.marker_cache.get(&(epoch_id, *object_id)) {
```

Comment:
I'm guessing that access to this function is locked in some way? Otherwise is there a possible race condition here (and elsewhere) where something could be added to the `dirty` between the query of it and the fallback to `cached` and `store`.

Reply:
that is a good point and I think i will fix this just to avoid any confusion, but i don't think this is an actual issue, because in either this or the fully-locked case, you can always miss an insert to the dirty set that you are racing with.